### PR TITLE
feature: http interactions preparations

### PR DESCRIPTION
**Pull request by l7ssha**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Use smart commits here to manipulate issues (eg. Fixes #issue)

## Connected issues & other potential problems

If changes in PR are connected to other issues or are affecting code in other parts of framework
(e.g. in main package or any other subpackage) make sure to link issue/PR and describe said problem 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [ ] Ran `dart analyze` or `make analyze` and fixed all issues
- [ ] Ran `dart format --set-exit-if-changed -l 160 ./lib` or `make format` and fixed all issues
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/src/client.dart [hunk 1]
```diff
@@ -132,7 +132,7 @@ abstract class Nyxx {
 }
 
 /// A client that can make requests to the HTTP API and is authenticated with a bot token.
-class NyxxRest with ManagerMixin implements Nyxx {
+class NyxxRest with ManagerMixin, EventRestMixin implements Nyxx {
```
Comment (abitofevrything) [inline lib/src/client.dart:135]: I'd rather not all `NyxxRest` clients have the `onInteractionCreate` stream as most HTTP only clients will not be receiving the events.

If the events are only emitted if a plugin is added to the client, then the stream should be on the plugin itself.

File: lib/src/client.dart [hunk 2]
```diff
@@ -180,7 +180,7 @@ class NyxxRest with ManagerMixin implements Nyxx {
   }
 }
 
-class NyxxOAuth2 with ManagerMixin implements NyxxRest {
+class NyxxOAuth2 with ManagerMixin, EventRestMixin implements NyxxRest {
```
Comment (abitofevrything) [inline lib/src/client.dart:183]: The above comment also solves the issue here as OAuth2 clients _cannot_ receive interaction events (and also cannot connect to the gateway, which is why NyxxOAuth2 doesn't implement NyxxGateway).

File: lib/src/event_mixin.dart [hunk 3]
```diff
@@ -18,10 +19,21 @@ import 'package:nyxx/src/models/gateway/events/webhook.dart';
 import 'package:nyxx/src/models/interaction.dart';
 import 'package:nyxx/src/utils/iterable_extension.dart';
 
+mixin EventRestMixin implements Nyxx {
+  @internal
+  StreamController<Event> onEventController = StreamController.broadcast();
```
Comment (abitofevrything) [inline lib/src/event_mixin.dart:24]: We should avoid exposing controllers as part of the API. Events should be added to a stream in the same place the stream is created, otherwise the code becomes hard to follow.

Moving the stream to the plugin avoids this problem.

File: lib/src/gateway/shard.dart [hunk 4]
```diff
@@ -52,7 +52,7 @@ class Shard extends Stream<ShardMessage> implements StreamSink<GatewayMessage> {
       } else if (message is EventReceived) {
         final event = message.event;
 
-        if (event is! RawDispatchEvent) {
+        if (event is GatewayEvent) {
```
Comment (abitofevrything) [inline lib/src/gateway/shard.dart:55]: Changing this line modifies the logic so it is now incorrect - `event` is always a `GatewayEvent`, so the `else` branch that outputs more information for dispatch events is always skipped, which reduces verbose logging output.

File: lib/src/gateway/message.dart [hunk 5]
```diff
@@ -34,7 +34,7 @@ abstract class ShardMessage with ToStringHelper {}
 /// A shard message sent when an event is received on the Gateway.
 class EventReceived extends ShardMessage {
   /// The event that was received.
-  final GatewayEvent event;
+  final Event event;
```
Comment (abitofevrything) [inline lib/src/gateway/message.dart:37]: Shards only ever emit `GatewayEvent`s, so this can be changed back.

File: lib/src/models/gateway/event.dart [hunk 6]
```diff
@@ -5,11 +5,12 @@ import 'package:nyxx/src/utils/to_string_helper/to_string_helper.dart';
 /// {@template gateway_event}
 /// The base class for all events received from the Gateway.
 /// {@endtemplate}
-abstract class GatewayEvent with ToStringHelper {
+abstract class Event with ToStringHelper {}
+
+abstract class GatewayEvent extends Event {
```
Comment (abitofevrything) [inline lib/src/models/gateway/event.dart:10]:
```suggestion
/// The base class for all events received by a client.
abstract class Event with ToStringHelper {}

/// {@template gateway_event}
/// The base class for all events received from the Gateway.
/// {@endtemplate}
abstract class GatewayEvent extends Event {
````

File: lib/src/models/gateway/event.dart [hunk 7]
```diff
@@ -5,11 +5,12 @@ import 'package:nyxx/src/utils/to_string_helper/to_string_helper.dart';
...
   /// The opcode of this event.
   final Opcode opcode;
 
-  /// {@macro gateway_event}
```
Comment (abitofevrything) [inline lib/src/models/gateway/event.dart:12]: Add this line back once the documentation is fixed.

File: lib/src/models/gateway/events/interaction.dart [hunk 8]
```diff
@@ -4,10 +4,10 @@ import 'package:nyxx/src/models/interaction.dart';
 /// {@template interaction_create_event}
 /// Emitted when an interaction is received by the client.
 /// {@endtemplate}
-class InteractionCreateEvent<T extends Interaction<dynamic>> extends DispatchEvent {
+class InteractionCreateEvent<T extends Interaction<dynamic>> extends Event {
```
Comment (abitofevrything) [inline lib/src/models/gateway/events/interaction.dart:7]: There isn't a need for this change, a cleaner solution would simply be to have another type of interaction event emitted when using HTTP interactions.

For example, `InteractionCreateEvent` would only be emitted over the gateway, and `WebhookInteractionCreateEvent` (which would not extend `GatewayEvent`) would be emitted when using HTTP interactions.

Both events would simply contain an `interaction` field (and could both implement a common interface), but keeping `InteractionCreateEvent` as a `GatewayEvent` allows for tighter typing in a lot of places (e.g `Gateway.onEvent` and `EventMixin.onEvent`) and allows users to access the client associated with an interaction using `event.gateway.client`,

**Review by abitofevrything — CHANGES_REQUESTED** [review]

File: lib/src/client.dart (hunk 1 above)
Comment (l7ssha) [reply lib/src/client.dart:135]: yeah,  stream could be added to plugin but then implementation for nyxx commands would be hard or impossible. We have to have some kind of support in main Nyxx interfaces. Also most client might not receive events but it's supported feature - I dont see problem with exposing possible feature even if rarely it would be used.
